### PR TITLE
hide Flag, by making it a class of static operations (GEN-554)

### PR DESCRIPTION
**Pull request by littleredcomputer**

### CODE REVIEW

File: src/genjax/_src/core/generative/choice_map.py [hunk 1]
```diff
@@ -153,7 +157,7 @@ def all() -> "Selection":
             from genjax import Selection
 
             all_selection = Selection.all()
-            assert all_selection["any_address"].f == True
+            assert all_selection["any_address"] == True
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:185]: I like it so far!

File: src/genjax/_src/core/generative/choice_map.py [hunk 2]
```diff
@@ -321,12 +325,12 @@ class AllSel(Selection):
...
         ```python exec="yes" html="true" source="material-block" session="choicemap"
         all_sel = Selection.all()
-        assert all_sel["any_address"].f == True
+        assert all_sel["any_address"] == True
         ```
     """
 
     def check(self) -> Flag:
-        return Flag(True)
+        return True
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py:352]: I can pull this into my PR easily, so good

File: src/genjax/_src/core/generative/choice_map.py [hunk 3]
```diff
@@ -1268,16 +1266,16 @@ def build(
...
         staged_err(
             err_check,
             f"The disjoint union of two choice maps have a value collision:\nc1 = {self.c1}\nc2 = {self.c2}",
         )
         v1 = self.c1.get_value()
         v2 = self.c2.get_value()
 
-        def pair_flag_to_idx(first: Flag, second: Flag):
-            return first.f + 2 * second.f - 1
+        def pair_flag_to_idx(first, second):
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py]: hey, where my types at

File: src/genjax/_src/core/generative/choice_map.py [hunk 4]
```diff
@@ -1389,9 +1387,9 @@ def build(
         c: ChoiceMap,
         flag: Flag,
     ) -> ChoiceMap:
-        if c.static_is_empty() | flag.concrete_true():
+        if c.static_is_empty() | FlagOp.concrete_true(flag):
             return c
-        elif flag.concrete_false():
+        elif FlagOp.concrete_false(flag):
```
Comment (sritchie) [inline src/genjax/_src/core/generative/choice_map.py]: nice, these are the right checks for concrete, I like it. I had wanted `is_concrete` as well...

File: src/genjax/_src/core/generative/functional_types.py [hunk 5]
```diff
@@ -57,22 +59,23 @@ class Mask(Generic[R], Pytree):
     If you use invalid `Mask(False, data)` data in inference computations, you may encounter silently incorrect results.
     """
 
-    flag: Flag
+    flag: Flag | Diff[Flag]
     value: R
 
     @staticmethod
-    def maybe(f: Flag, v: "R | Mask[R]") -> "Mask[R]":
+    def maybe(f: Flag | Diff[Flag], v: "R | Mask[R]") -> "Mask[R]":
```
Comment (sritchie) [inline src/genjax/_src/core/generative/functional_types.py:76]: you could provide an `@overload` that shows that if a Mask comes in we'd better not see a diff?

File: src/genjax/_src/core/generative/functional_types.py [hunk 6]
```diff
@@ -112,10 +115,17 @@ def _check():
...
 
             def inner(true_v: ArrayLike, false_v: ArrayLike) -> Array:
-                return jnp.where(self.flag.f, true_v, false_v)
+                return jnp.where(self.primal_flag(), true_v, false_v)
 
             return jtu.tree_map(inner, self.value, default)
 
+    def primal_flag(self) -> Flag:
+        match self.flag:
+            case Diff(primal, _):
```
Comment (sritchie) [inline src/genjax/_src/core/generative/functional_types.py:134]: maybe `@property` here?

File: src/genjax/_src/core/typing.py [hunk 7]
```diff
@@ -52,6 +52,7 @@
 Int = int
 Float = float
 Bool = bool
+Flag = Bool | BoolArray
```
Comment (sritchie) [inline src/genjax/_src/core/typing.py:55]: woohoo, look at that!

File: src/genjax/_src/generative_functions/combinators/switch.py [hunk 8]
```diff
@@ -84,7 +84,7 @@ def get_sample(self) -> Sample:
             chm = ChoiceMap.empty()
             for _idx, _chm in enumerate(subsamples):
                 assert isinstance(_chm, ChoiceMap)
-                masked_submap = _chm.mask(Flag(jnp.all(_idx == idx)))
+                masked_submap = _chm.mask(jnp.all(_idx == idx))
```
Comment (sritchie) [inline src/genjax/_src/generative_functions/combinators/switch.py]: my PR kills this jnp.all, I don't think we want this as it prevents `_idx == idx` from staying concrete.

File: src/genjax/_src/core/interpreters/staging.py [hunk 9]
```diff
@@ -39,89 +39,80 @@
 # Concrete Boolean arithmetic #
 ###############################
 
+R = TypeVar("R")
 
-@Pytree.dataclass
-class Flag(Pytree):
+
+class FlagOp:
```
Comment (sritchie) [inline src/genjax/_src/core/interpreters/staging.py:45]: cool, I like it. The name is still a little weird. JAXBool or something might be clearer, but everyone is used to Flag so no need to change.

**Review by sritchie — APPROVED** [review]
Nice work, so good to see the types here get cleaned up. This simplifies all of the assertion in choice_map, such a great sign.

File: src/genjax/_src/core/generative/choice_map.py (hunk 2 above)
Comment (sritchie) [reply src/genjax/_src/core/generative/choice_map.py:352]: do we want this to return `jnp.asarray(True)`? should `check` return `BoolArray`, to match the jax "ArrayLike in, Array out"?

File: src/genjax/_src/core/interpreters/staging.py (hunk 9 above)
Comment (sritchie) [reply src/genjax/_src/core/interpreters/staging.py:45]: What about `BoolArrayLike`? that matches the pattern... I suspect you are not going to like this suggestion.

**Comment by sritchie** [comment]
I'll close #1314 ... can you merge that into here?

File: src/genjax/_src/core/generative/choice_map.py (hunk 2 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/generative/choice_map.py:352]: My gut says that we should return a concrete value when we know we can (and this is the point of the type AllSel, a selection that is known to match at compile time)

File: src/genjax/_src/core/generative/choice_map.py (hunk 3 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/generative/choice_map.py]: not sure! they're back now

File: src/genjax/_src/core/generative/functional_types.py (hunk 5 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/generative/functional_types.py:76]: I tried this:
```
    @overload
    @staticmethod
    def maybe(f: Flag, v: "Mask[R]") -> "Mask[R]":
        ...

    @overload
    @staticmethod
    def maybe(f: Flag | Diff[Flag], v: R) -> "Mask[R]":
        ...
```
but it did not redline the following test invocation:
```
Mask.maybe(Diff.no_change(True), Mask(True, 99))
```
Because the Mask type matches R in the second  overload :(

File: src/genjax/_src/core/generative/functional_types.py (hunk 6 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/generative/functional_types.py:134]: Hm. Maybe the thing to do is call the property `.flag` and rename the actual inner value? The argument would be that nobody is ever interested in the Diff form of a mask: it has to be that way to satisfy the diff-leaf property but has no meaning beyond that

File: src/genjax/_src/generative_functions/combinators/switch.py (hunk 8 above)
Comment (littleredcomputer) [reply src/genjax/_src/generative_functions/combinators/switch.py]: Good point, removing it also

File: src/genjax/_src/core/interpreters/staging.py (hunk 9 above)
Comment (littleredcomputer) [reply src/genjax/_src/core/interpreters/staging.py:45]: The alternatives are not bad but none seems better that FlagOp to me atm

**Comment by sritchie** [comment]
Brilliant!!

**Comment by sritchie** [comment]
@littleredcomputer I went ahead and linked this to the Linear issue. nice work!
